torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,394 @@
+
+            # Update the scale for next iteration.
+            scaler.update()
+
+    See the :ref:`Gradient Scaling Examples<gradient-scaling-examples>` for usage in more complex cases.
+
+    ``scaler`` maintains the scale factor internally.  To leverage ``float16``'s full dynamic range,
+    ``scaler`` attempts to use the largest scale factor it can without incurring overflow.
+    It does so by checking the gradients for infs and NaNs during every :meth:`step` or separate :meth:`unscale`.
+    If no infs/NaNs are found, :meth:`step` runs the underlying ``optimizer.step()`` as usual and
+    :meth:`update` multiplies the scale factor by the growth factor.  If infs/NaNs are found, :meth:`step` skips the
+    underlying ``optimizer.step()`` (so the params themselves remain uncorrupted) and multiplies the scale factor by
+    the backoff factor.
+
+    The scale factor often causes infs/NaNs to appear in gradients for the first few iterations as its
+    value calibrates.  ``scaler.step`` will skip the underlying ``optimizer.step()`` for these
+    iterations.  After that, step skipping should occur rarely (once every few hundred iterations).
+
+    Arguments:
+        init_scale (float, optional, default=2.**24):  Initial scale factor.
+        growth_factor (float, optional, default=1.001):  Factor by which the scale is multiplied during
+            :meth:`update` if no inf/NaN gradients were found this iteration.
+        backoff_factor (float, optional, default=0.5):  Factor by which the scale is multiplied during
+            :meth:`update` if inf/NaN gradients were found this iteration.
+        enabled (bool,optional, default=True):  If ``False``, disables gradient scaling. :meth:`step` simply
+            invokes the underlying ``optimizer.step()``, and other methods become no-ops.
+    """
+    def __init__(self,
+                 init_scale=2.**24,
+                 growth_factor=1.001,
+                 backoff_factor=0.5,
+                 enabled=True):
+        self._enabled = enabled
+        if enabled:
+            self._init_scale = init_scale
+            # self._scale will be lazily initialized during the first call to scaler.scale(loss or outputs)
+            self._scale = None
+            self._growth_factor = growth_factor
+            self._backoff_factor = backoff_factor
+            self._per_optimizer_states = defaultdict(dict)
+
+    @staticmethod
+    def _scale_not_initialized_error(funcname):
+        return "Attempted to call {} but the scale tensor is None. This may indicate your ".format(funcname) + \
+               "script did not use scaler.scale(loss or outputs) earlier in the iteration."
+
+    def scale(self, outputs):
+        """
+        Multiplies ('scales') a tensor or list of tensors by the scale factor.
+
+        These tensors are typically the outputs of a network. When ``backward()``
+        is called on scaled outputs the resulting gradients will also be scaled.
```

Comment:
... this will cause the gradients created from loss.backward() to be scaled by the same factor.